### PR TITLE
Add in security reporting

### PR DESCRIPTION
**Pull request by samhotchkiss**

### CODE REVIEW

**Comment by samhotchkiss** [comment]
Hey @georgestephanis -- after your last feedback, I got to thinking about this, and built it out as a quasi-hybrid of the first two methods.  We use a hook to "call" for updates, but I wasn't comfortable just letting everybody filter on one array, too much possibility for one plugin to make a slight mistake and wipe out all of our data.

So, now it puts out the action, then plugins can call in their reports with something like this:
```php
	function akismet_submit_jetpack_security_report() {
		Jetpack::submit_security_report( 'spam', 'akismet', $args = array( 'plugin' => 'Akismet', 'blocked' => 138284 ) );
	}
	add_action( 'jetpack_security_report', 'akismet_submit_jetpack_security_report' );
```

We then take the data they submit, sanitize it, compile it, and sync it back to wpcom.

Thoughts on this approach?  (I know I need to build out docblocs, but any other code style feedback is welcome, too)

File: class.jetpack.php [hunk 1]
```diff
@@ -1041,6 +1049,81 @@ public function check_twitter_tags() {
...
+			$args[ 'blocked' ] = (int)$args[ 'blocked' ];
+		}
+		if( isset( $args[ 'last' ] ) ) {
+			$args[ 'last' ] = (int)$args[ 'last' ];
+		}
+		if( isset( $args[ 'next' ] ) ) {
+			$args[ 'next' ] = (int)$args[ 'next' ];
+		}
+		
+		$this->security_report[$type][$plugin_slug] = $args;
```
Comment (georgestephanis) [inline class.jetpack.php]: whitespace inside brackets around variables -- strings don't get whitespace, but variables should.

File: class.jetpack.php [hunk 2]
```diff
@@ -1041,6 +1049,81 @@ public function check_twitter_tags() {
...
+		// Sanitize everything to make sure we're not syncing something wonky
+		$type = sanitize_key( $type );
+		$plugin_slug = sanitize_key( $plugin_slug );
+		
+		$args[ 'plugin' ] = sanitize_title( $args[ 'plugin' ] );
+		
+		// Cast blocked, last and next as integers.
+		// Last and next should be in unix timestamp format
+		if( isset( $args[ 'blocked' ] ) ) {
+			$args[ 'blocked' ] = (int)$args[ 'blocked' ];
```
Comment (georgestephanis) [inline class.jetpack.php]: Should also have a space between the cast and the variable.

File: class.jetpack.php [hunk 3]
```diff
@@ -1041,6 +1049,81 @@ public function check_twitter_tags() {
...
+		
+		// Sanitize everything to make sure we're not syncing something wonky
+		$type = sanitize_key( $type );
+		$plugin_slug = sanitize_key( $plugin_slug );
+		
+		$args[ 'plugin' ] = sanitize_title( $args[ 'plugin' ] );
+		
+		// Cast blocked, last and next as integers.
+		// Last and next should be in unix timestamp format
+		if( isset( $args[ 'blocked' ] ) ) {
```
Comment (georgestephanis) [inline class.jetpack.php]: space between if and parens

**Comment by georgestephanis** [comment]
Honestly, I think you're still overthinking it.

Think of it like how core handles plugin update api calls.

`_maybe_update_plugins()` is a wrapper that will either return a transient, or call a function that hits the api and refresh the transient.

https://github.com/WordPress/WordPress/blob/fc843ce4d0147d1851fa2e64ae7184f59a1b658b/wp-includes/update.php#L589-L604

I'd say we should do something like ..

```php
function get_jetpack_security_details() {
	$current = Jetpack_Options::get_option( 'security_details', array() );

	if ( isset( $current['last_checked'] ) && 12 * HOUR_IN_SECONDS > ( time() - $current['last_checked'] ) ) {
		return $current['details'];
	}

	$new = array(
		'last_checked' => time(),
		'details' => fetch_jetpack_security_details(),
	);
	Jetpack_Options::update_option( 'security_details', $new );

	return $new['details'];
}

function fetch_jetpack_security_details() {
	$details = array();

	$details['spam']    = apply_filters( 'jetpack_security_details_spam',    array() );
	$details['backup']  = apply_filters( 'jetpack_security_details_backup',  array() );
	$details['login']   = apply_filters( 'jetpack_security_details_login',   array() );
	$details['malware'] = apply_filters( 'jetpack_security_details_malware', array() );
	
	return $details;
}
```

I really, really feel it should be polling each provider for updates, rather that expecting them to push things to us.  It's far less intrusive that way.

**Comment by samhotchkiss** [comment]
With my method, we are polling them for updates, but in a structured way.  Every X minutes/hours, we fire the `jetpack_security_report` action, and they submit their data-- it only differs from offering them a filter because it allows us to sanitize the data and prevent them from accidentally squashing the reports of others.

In your proposal, there's no format for the data, so what do we do with it when we get it?  If there are multiple backup providers, how to we keep them from stepping on each others toes?

**Comment by georgestephanis** [comment]
As it's an array being passed around, each uses a key to pass their data in.  So the result is an associative array of provider => data and we can squash it as we need to.

As for the data format, that's still up in the air.  We do need to get that sorte,d but it will be dependent on the output ui.  I'd roughed in a good bit over on the `security` branch for output and schema.  https://github.com/Automattic/jetpack/compare/security

**Comment by samhotchkiss** [comment]
That's what I'm saying-- we're passing around an array that would make it very easy to accidentally disturb another plugin's data.  It seems too delicate.  I guess I'm not seeing how the alternative I built out is making things harder on anybody-- it's easy to get your data in, it's easy to know how to pass it, you don't have to worry about conflicts, and we're able to sanitize everything so it'll display nicely in calypso and in the plugin.

The only data I saw over in `security` is slug and label, we're definitely going to need more than that based on the display mockups.

**Comment by samhotchkiss** [comment]
@georgestephanis, per our chat, here's an idea of what the collected data should look like:  (as php instead of json for readability)

```php
	array(
		'type'        => 'spam',
		'plugin'      => 'akismet', // Plugin slug
		'plugin_name' => 'Akismet', // Plugin name
		'blocked'     => 1827329, // (int) number of blocked spam or attempts
		'status'      => 'ok' // Status (ok, warning, error)
	);
		
	array(
		'type'        => 'spam',
		'plugin'      => 'akismet', // Plugin slug
		'plugin_name' => 'Akismet', // Plugin name
		'blocked'     => 1827329, // (int) number of blocked spam or attempts
		'status'      => 'error', // Status (ok, warning, error)
		'message'     => 'Subscription Expired' // String message if status is warning or error.  A, em, and strong tags allowed, all others stripped
	);
	
	array(
		'type'        => 'backup',
		'plugin'      => 'backupbuddy', // Plugin slug
		'plugin_name' => 'BackupBuddy', // Plugin name
		'last'        => 1422038224, // Unix timestamp of last time backup occurred
		'next'        => 1422058374, // Unix timestamp of next scheduled backup
		'status'      => 'warning', // Status (ok, warning, error)
		'message'     => 'Allocated storage 90% full, please <a href="http://backupbuddy.com/my-account/">purchase more storage</a>.' // String message if status is warning or error.  A, em, and strong tags allowed, all others stripped
	);
```

**Comment by georgestephanis** [comment]
Okay, well should we maybe have these be instantiations of a specific class -- so they're objects rather than just arrays, to better normalize the data?  What types of statuses can there be specifically -- I see spam and backup, what others?  I think the plugin slug itself is probably sufficient, we can pull the name ourselves if we need it.

**Comment by samhotchkiss** [comment]
@georgestephanis:

Available types are `spam` `backup` `botnet` `file_scanning`

Normalization was the goal of the `submit_security_report` function (still need to add status and message), since it's all going to get wrapped up and sent to wpcom, where it will actually be displayed (there's not going to be a dashboard in the plugin aside from the widget).

Do we have a good way of getting the plugin name on wpcom?

**Comment by georgestephanis** [comment]
`botnet` I'd think may be better as something specifying securing your login form.  Some folks may prefer to use limit login attempts or a two-factor authentication plugin for those purposes.  `file_scanning` may also be better as `malware` -- unsure.

I'd prefer to have Jetpack fetch the plugin name, to correct for any variances of the data being passed in.  Perhaps just pass in the plugin path, like `jetpack/jetpack.php` or the like.

File: class.jetpack.php [hunk 4]
```diff
@@ -1041,6 +1049,103 @@ public function check_twitter_tags() {
...
+			return new Jetpack_Error( 'invalid_security_report', 'Invalid Security Report' );
+		}
+		
+		//Get rid of any non-allowed args
+		$args = array_intersect_key( $args, array( 'blocked', 'last', 'next', 'status', 'message' ) );
+		
+		$plugin = get_plugin_data( $plugin_file );
+		
+		if ( !$plugin['Name'] ) {
+			return new Jetpack_Error( 'security_report_missing_plugin_name', 'Invalid Plugin File Provided' );
```
Comment (georgestephanis) [inline class.jetpack.php]: Why is this a Jetpack_Error, not WP_Error?

Comment (samhotchkiss) [inline class.jetpack.php]: Because it's within Jetpack and Jetpack_Error is a wrapper for WP_Error, so I figured that was the preferred way to handle

**Comment by samhotchkiss** [comment]
Feeling pretty good about this in its current state, and tested, but I'd love to get some more eyes before I merge it in, could you take a look @enejb @roccotripaldi @dereksmart ?

**Pull request merged**